onnxruntime/core/providers/cpu/math/element_wise_ops.cc

```diff
@@ -217,6 +219,43 @@ Status Sub<T>::Compute(OpKernelContext* context) const {
       [](EigenVectorMap<T> output, ConstEigenVectorMap<T> input0, ConstEigenVectorMap<T> input1) { output = input0 - input1; });
 }
 
+template <>
+Status Sub<MLFloat16>::Compute(OpKernelContext* context) const {
+  TBroadcaster<MLFloat16, MLFloat16> mod_broadcaster{*context->Input<Tensor>(0), *context->Input<Tensor>(1)};
+  Tensor* output = context->Output(0, mod_broadcaster.GetOutputShape());
+  TBroadcastOutput<MLFloat16> mod_broadcast_output{
+      mod_broadcaster.GetSpanSize(), *output};
+  BroadcastLoopSpan(
+      mod_broadcaster, mod_broadcast_output,
+      [](gsl::span<MLFloat16> output, const MLFloat16& X, gsl::span<const MLFloat16> Y) {
+        auto x_data = Eigen::half(X.val);
+        auto y_data = Eigen::Map<const Eigen::Matrix<Eigen::half, Eigen::Dynamic, 1>>(
+            static_cast<const Eigen::half*>(static_cast<const void*>(Y.data())), Y.size());
```

Comment:
>static_cast [](start = 44, length = 11)

reinterpret_cast<const Eigen::half*>() see other cases for float16 support You do not need two casts